helix-core/src/main/java/org/apache/helix/task/TaskUtil.java

```diff
@@ -733,10 +734,29 @@ protected static boolean removeJobsFromWorkflow(final HelixDataAccessor dataAcce
     if (workflowContext != null) {
       Map<String, TaskState> jobStates = workflowContext.getJobStates();
       for (String job : workflowConfig.getJobDag().getAllNodes()) {
+        if (expiredJobs.contains(job)) {
+          continue;
+        }
         JobConfig jobConfig = TaskUtil.getJobConfig(dataAccessor, job);
         JobContext jobContext = TaskUtil.getJobContext(propertyStore, job);
-        if (isJobExpired(job, jobConfig, jobContext, jobStates.get(job))) {
+        TaskState jobState = jobStates.get(job);
```

Comment:
Can you confirm if jobState here is not null always?

Reply:
jobState could be null in an erroneous state where a job is in the DAG but not in jobState map, but is that a situation we need to consider? This logic here wouldn't interact with such a job as it fails the jobState check conditions. I think it's fine to not deal with that case.